File: sources/scalac/transformer/matching/CodeFactory.java
/*     ____ ____  ____ ____  ______                                     *\
**    / __// __ \/ __// __ \/ ____/    SOcos COmpiles Scala             **
**  __\_ \/ /_/ / /__/ /_/ /\_ \       (c) 2002, LAMP/EPFL              **
** /_____/\____/\___/\____/____/                                        **
**                                                                      **
** $Id$
\*                                                                      */

package scalac.transformer.matching;

import ch.epfl.lamp.util.Position;

import scalac.*;
import scalac.ast.*;
import scalac.util.*;
import scalac.symtab.*;
import scalac.typechecker.*;
import PatternNode.*;
import Tree.*;

class CodeFactory extends PatternTool {

    public int pos = Position.FIRSTPOS ;

    static final Name TUPLE2_N       = Name.fromString("scala.Tuple2");

    static final Name LEFT_N = Name.fromString("_1");
    static final Name RIGHT_N = Name.fromString("_2");
    static final Name HEAD_N       = Name.fromString("head");
    static final Name TAIL_N       = Name.fromString("tail");
    static final Name ISEMPTY_N        = Name.fromString("isEmpty");

    Symbol refSym() { // delete Names.Ref
	return defs.getType( Names.scala_Ref ).symbol() ;
    }

    Symbol seqSym() {
	return defs.getType( Names.scala_Seq ).symbol() ;
    }

    Symbol intSym() {
	return defs.getType( Names.scala_Int ).symbol() ;
    }

    Symbol tuple2Sym() {
	return defs.getType( TUPLE2_N ).symbol() ;
    }

    Symbol iteratorSym() {
	return defs.getType( Names.scala_Iterator ).symbol() ;
    }

    Symbol seqListSym() {
	return defs.getType( Names.scala_List ).symbol() ;
    }

    Symbol seqListSym_isEmpty() {
	return seqListSym().lookup( ISEMPTY_N );
    }

    Symbol seqListSym_head() {
	return seqListSym().lookup( HEAD_N );
    }

    Symbol seqListSym_tail() {
	return seqListSym().lookup( TAIL_N );
    }

    /*
    Symbol seqConsSym() {
	return defs.getType( Names.scala_COLONCOLON ).symbol() ;
    }

    Symbol seqNilSym() {
	return defs.getType( Names.scala_Nil ).symbol().module(); // no need for TypeApply anymore!x
    }
    */
    Symbol seqIterSym() {
	return defs.getType( Names.scala_Iterator ).symbol();
    }

    Symbol seqIterSym_next() {
	Scope scp = seqIterSym().members(); return scp.lookup( Names.next );
    }

    Symbol seqIterSym_hasNext() {
	Scope scp = seqIterSym().members(); return scp.lookup( Names.hasNext );
    }
    /*

    Symbol seqTraceSym() {
	return defs.getType( Name.fromString( "scala.SeqTrace" ) ).symbol();
    }
    Symbol seqTraceConsSym() {
	return defs.getType( Name.fromString( "scala.SeqTraceCons" ) ).symbol();
    }

    Symbol seqTraceNilSym() {
	return defs.getType( Name.fromString( "scala.SeqTraceNil" ) ).symbol();
    }
    */
    Symbol iterableSym() {
	return defs.getType( Names.scala_Iterable ).symbol();
    }

    Symbol newIterSym() {
	Scope scp = iterableSym().members(); return scp.lookup( Names.elements );
    }

    Symbol andSym() {
        return defs.BOOLEAN_CLASS.lookup(AND_N);
    }

    Symbol orSym() {
        return defs.BOOLEAN_CLASS.lookup(OR_N);
    }

    Symbol failSym() {
        return defs.SCALA_CLASS.lookup(MATCHERROR_N).lookup(FAIL_N);
    }


    public CodeFactory( Unit unit, Infer infer, int pos ) {
	super( unit, infer );
	this.pos = pos;
    }

    // --------- these are new

    /** If  ... pos, type is copied from thenBody
     */
    Tree If( Tree cond, Tree thenBody, Tree elseBody ) {
	assert cond != null:"cond is null";
	assert thenBody != null:"thenBody is null";
	assert elseBody != null:"elseBody is null";
	return gen.If( thenBody.pos, cond, thenBody, elseBody );
    }

    /** a faked switch statement
     */
    Tree Switch( Tree selector,
		 Tree condition[],
		 Tree body[],
		 Tree defaultBody ) {
	assert selector != null:"selector is null";
	assert condition != null:"cond is null";
	assert body != null:"body is null";
	assert defaultBody != null:"defaultBody is null";
	Tree result = defaultBody;

	for( int i = condition.length-1; i >= 0; i-- )
	    result = If( condition[ i ],
			 body[ i ],
			 result
			 ).setType( result.type );

	return result ;
    }

    /** returns `List[ elemType ]' */
      Type SeqListType( Type elemType ) {
            return Type.TypeRef( defs.SCALA_TYPE,
                                 seqListSym(),
                                 new Type[] { elemType });
      }

    /** returns  `List[ Tuple2[ scala.Int, <elemType> ] ]' */
      Type SeqTraceType( Type elemType ) {
	  Type t = Type.TypeRef( defs.SCALA_TYPE,
                                 seqListSym(),
                                 new Type[] { pairType( defs.INT_TYPE,
							elemType ) });
	  //System.err.println("CodeFactory::SeqTraceType -"+ t );
	  return t;
      }

    /**  returns `Iterator[ elemType ]' */
    Type _seqIterType( Type elemType ) {
	Symbol seqIterSym = defs.getType( Names.scala_Iterator ).symbol();

	return Type.TypeRef( defs.SCALA_TYPE, seqIterSym(),
			     new Type[] { elemType });
    }

    /**  returns `<seqObj.elements>' */
    Tree newIterator( Tree seqObj, Type elemType ) {
	Symbol newIterSym = newIterSym();
	Tree t1 = gen.Select( pos, seqObj, newIterSym)
	    .setType( Type.MethodType(new Symbol[] {},_seqIterType( elemType )));

	Tree theIterator = gen.Apply(seqObj.pos,
				     t1,
				     Tree.EMPTY_ARRAY)
	    .setType( _seqIterType( elemType ) );

	return theIterator;

    }

    /** returns code `<seqObj>.elements'
     *  the parameter needs to have type attribute `Sequence[<elemType>]'
     */
    Tree newIterator( Tree seqObj ) {
	return newIterator( seqObj, getElemType_Sequence( seqObj.type() ));
    }

    /** code `Nil'
        Tree _seqTraceNil( Type elemType ) {
	return newSeqNil( null );
    }
     */

      //                       `SeqCons[ elemType ]'
    /*

      Type _seqConsType( Type elemType ) {
            return Type.TypeRef( defs.SCALA_TYPE,
                                 seqConsSym(),
                                 new Type[] { elemType });
      }

    Tree newSeqNil( Type tpe ) {
	return gen.Select(gen.Ident(pos, defs.SCALA), seqNilSym());
    }
    */

    // EXPERIMENTAL
    Tree newRef( Tree init ) {
	//System.out.println( "hello:"+refSym().type() );
	return gen.New(gen.mkPrimaryConstr(pos, refSym(),
			                    new Type[] { init.type() },
			                    new Tree[] { init } ));
    }

    /** returns A for T <: Sequence[ A ]
     */
    Type getElemType_Sequence( Type tpe ) {
	//System.err.println("getElemType_Sequence("+tpe.widen()+")");
	Type tpe1 = tpe.widen().baseType( seqSym() );

	if( tpe1 == Type.NoType )
	    throw new ApplicationError("arg "+tpe+" not subtype of Sequence[ A ]");

	return tpe1.typeArgs()[ 0 ];
    }

    /** returns A for T <: Iterator[ A ]
     */
    Type getElemType_Iterator( Type tpe ) {
	//System.err.println("getElemType_Iterator("+tpe+")");

	Type tpe1 = tpe.widen().baseType( iteratorSym() );

	switch( tpe1 ) {
	case TypeRef(_,_,Type[] args):
	    return args[ 0 ];
	default:
	    throw new ApplicationError("arg "+tpe+" not subtype of Iterator[ A ]");
	}

    }

    /** `it.next()'
     */
    public Tree _next( Tree iter ) {
	return gen.mkApply__(gen.Select(iter, seqIterSym_next()));
    }

    /** `it.hasNext()'
     */
    public Tree _hasNext( Tree iter ) {
	return gen.mkApply__(gen.Select(iter, seqIterSym_hasNext()));
    }

    /** `!it.hasCur()'
     */
    public Tree _not_hasNext( Tree iter ) {
	return gen.mkApply__(gen.Select(_hasNext(iter), notSym));
    }

      /** `trace.isEmpty'
       */
      public Tree isEmpty( Tree iter ) {
          return gen.mkApply__(gen.Select(iter, seqListSym_isEmpty()));
      }

    Tree SeqTrace_headElem( Tree arg ) { // REMOVE SeqTrace
	Scope scp = seqListSym().members();
	Symbol headSym = scp.lookup ( HEAD_N );
	assert headSym != Symbol.NONE;
	scp = tuple2Sym().members();
	Symbol leftSym = scp.lookup ( RIGHT_N );
	assert leftSym != Symbol.NONE;
	Tree t =  gen.Apply( gen.Select( pos, arg, headSym ),
			     Tree.EMPTY_ARRAY );
	return gen.Apply( gen.Select( pos, t, leftSym ),
			  Tree.EMPTY_ARRAY );
    }

    Tree SeqTrace_headState( Tree arg ) { // REMOVE SeqTrace
	Scope scp = seqListSym().members();
	Symbol headSym = scp.lookup ( HEAD_N );
	assert headSym != Symbol.NONE;
	scp = tuple2Sym().members();
	Symbol leftSym = scp.lookup ( LEFT_N );
	assert leftSym != Symbol.NONE;
	Tree t =  gen.Apply( gen.Select( pos, arg, headSym ),
			     Tree.EMPTY_ARRAY );
	t =  gen.Apply( gen.Select( pos, t, leftSym ),
			Tree.EMPTY_ARRAY );
	return t;

    }

    Tree SeqTrace_tail( Tree arg ) {
	Scope scp = seqListSym()/*seqTraceSym()*/.members(); // REMOVE SeqTrace
	Symbol tailSym = scp.lookup ( TAIL_N );
	assert tailSym != Symbol.NONE;
	//System.err.println( "CodeFactory::SeqTrace_tail " + tailSym +" "+ tailSym.type() );
	return gen.Apply( gen.Select( pos, arg, tailSym ),
			  Tree.EMPTY_ARRAY );
    }

    /** `<seqlist>.head()'
     */
    Tree SeqList_head( Tree arg ) {
	Scope scp = seqListSym().members();
	Symbol headSym = scp.lookup ( HEAD_N );
	assert headSym != Symbol.NONE;
	return gen.Apply( gen.Select( pos, arg, headSym ),
			  Tree.EMPTY_ARRAY );
    }

    /** return the analyzed type
     */
    public Type typeOf(Symbol sym) {
        return sym.type();
        //return sym.typeAt(unit.global.ANALYZER_PHASE.id);
    }


    protected Tree Block(int pos, Tree[] ts, Type tpe) {
        if (ts.length == 1)
            return ts[0];
        else if (ts.length > 1)
            switch (ts[ts.length - 1]) {
	    case Block(Tree[] ts0):
		Tree[] ts1 = new Tree[ts0.length + ts.length - 1];
		System.arraycopy(ts, 0, ts1, 0, ts.length - 1);
		System.arraycopy(ts0, 0, ts1, ts.length - 1, ts0.length);
		return Block(pos, ts1, tpe);
            }
        return gen.Block(pos, ts);
    }

     // unused
       public Tree Negate(Tree tree) {
       switch (tree) {
       case Literal(Object value):
       return gen.mkBooleanLit(tree.pos, !((Boolean)value).booleanValue());
       }
       return gen.Apply(tree.pos, gen.Select(tree, notSym));
       }

    protected Tree And(Tree left, Tree right) {
        switch (left) {
	case Literal(Object value):
	    return ((Boolean)value).booleanValue() ? right : left;
        }
        switch (right) {
	case Literal(Object value):
	    if (((Boolean)value).booleanValue()) return left;
        }
        return gen.mkApply_V(gen.Select(left, andSym()), new Tree[]{right});
    }

    protected Tree Or(Tree left, Tree right) {
        switch (left) {
	case Literal(Object value):
	    return ((Boolean)value).booleanValue() ? left : right;
        }
        switch (right) {
	case Literal(Object value):
	    if (!((Boolean)value).booleanValue()) return left;
        }
        return gen.Apply(gen.Select(left, orSym()), new Tree[]{right});
    }

    protected Tree Equals(Tree left, Tree right) {
        Symbol fun = unit.global.definitions.EQEQ;
        return gen.mkApply_V(gen.Select(left, fun), new Tree[]{right});
    }

    protected Tree ThrowMatchError(int pos, Type type) {
        return gen.mkApplyTV(
			     gen.mkRef(pos, failSym()),
                             new Tree[]{gen.mkType(pos, type)},
                             new Tree[]{
                                 gen.mkStringLit(pos, unit.toString()),
                                 gen.mkIntLit(pos, Position.line(pos))
                             });
    }


    Type pairType( Type left, Type right ) {
	return Type.TypeRef( defs.SCALA_TYPE,
			     tuple2Sym() ,
			     new Type[] { left, right } );
    }

    Tree newPair( Tree left, Tree right ) {
 	return gen.New(gen.mkPrimaryConstr(pos, tuple2Sym(),
			                   new Type[] { left.type(), right.type() },
			                   new Tree[] { left, right }));

    }

}

File: sources/scala/tools/scalac/transformer/matching/CodeFactory.scala
/*     ____ ____  ____ ____  ______                                     *\
**    / __// __ \/ __// __ \/ ____/    SOcos COmpiles Scala             **
**  __\_ \/ /_/ / /__/ /_/ /\_ \       (c) 2002-2005, LAMP/EPFL         **
** /_____/\____/\___/\____/____/                                        **
**                                                                      **
** $Id$
\*                                                                      */


import scala.tools.util.Position;

import scalac._;
import scalac.ast._;
import scalac.atree.AConstant._;
import scalac.util._;
import scalac.symtab._;
//import PatternNode._;
import Tree._;
package scala.tools.scalac.transformer.matching {

class CodeFactory(val unit: CompilationUnit, pos1: Int) extends PatternTool(unit) {

    var pos = pos1;

    // --------- these are new

    /** a faked switch statement
     */
  def Switch(condition: Array[Tree], body: Array[Tree], defaultBody: Tree): Tree = {
    //assert condition != null:"cond is null";
    //assert body != null:"body is null";
    //assert defaultBody != null:"defaultBody is null";
    var result = defaultBody;

    var i = condition.length-1;
    while (i >= 0) {
      result = gen.If(condition(i), body(i), result);
      i = i - 1
    }

    return result ;
  }

  /** returns  `List[ Tuple2[ scala.Int, <elemType> ] ]' */
  def SeqTraceType( elemType: Type  ):  Type = {
    defs.LIST_TYPE(pairType(defs.int_TYPE(), elemType));
  }

  /**  returns `Iterator[ elemType ]' */
  def _seqIterType( elemType: Type  ):  Type = {
    defs.ITERATOR_TYPE(elemType);
  }

    /**  returns `<seqObj.elements>' */
  def newIterator( seqObj:Tree , elemType:Type  ):  Tree  = {
    gen.mkApply__(gen.Select(seqObj, defs.ITERABLE_ELEMENTS()));
  }

  /** returns code `<seqObj>.elements'
   *  the parameter needs to have type attribute `Sequence[<elemType>]'
   */
  def  newIterator( seqObj:Tree  ): Tree = {
    newIterator( seqObj, getElemType_Sequence( seqObj.getType() ));
  }

    // EXPERIMENTAL
  def newRef(init: Tree): Tree = {
    //System.out.println( "hello:"+refSym().type() );
    gen.New( gen.mkApplyTV( gen.mkPrimaryConstructorGlobalRef( pos, defs.REF_CLASS),
                           Predef.Array[Type] ( init.getType() ),
                           Predef.Array[Tree] ( init ) ));
  }

  /** returns A for T <: Sequence[ A ]
   */
  def getElemType_Sequence( tpe: Type):  Type = {
    //System.err.println("getElemType_Sequence("+tpe.widen()+")");
    val tpe1 = tpe.widen().baseType( defs.SEQ_CLASS );

    if( tpe1 == Type.NoType )
      throw new ApplicationError("arg "+tpe+" not subtype of Seq[ A ]");

    return tpe1.typeArgs()( 0 );
  }

    /** returns A for T <: Iterator[ A ]
     */
    def getElemType_Iterator( tpe: Type  ): Type = {
	//System.err.println("getElemType_Iterator("+tpe+")");

	val tpe1 = tpe.widen().baseType( defs.ITERATOR_CLASS );

	tpe1 match {
	case Type.TypeRef(_,_,args) =>
	    return args( 0 );
	case _ =>
	  throw new ApplicationError("arg "+tpe+" not subtype of Iterator[ A ]");
	}

    }

  /** `it.next()'
   */
  def  _next( iter: Tree  ):  Tree = {
    gen.mkApply__(gen.Select(iter, defs.ITERATOR_NEXT()));
  }

  /** `it.hasNext()'
   */
   def _hasNext(  iter: Tree ):  Tree =  {
     gen.mkApply__(gen.Select(iter, defs.ITERATOR_HASNEXT()));
   }

  /** `!it.hasCur()'
   */
  def  _not_hasNext( iter:Tree  ): Tree = {
    gen.mkApply__(gen.Select(_hasNext(iter), defs.BOOLEAN_NOT()));
  }

  /** `trace.isEmpty'
   */
  def isEmpty( iter: Tree  ):  Tree = {
    gen.mkApply__(gen.Select(iter, defs.LIST_ISEMPTY()));
  }

  def SeqTrace_headElem( arg: Tree  ):  Tree = { // REMOVE SeqTrace
    val t = gen.mkApply__(gen.Select(arg, defs.LIST_HEAD()));
    gen.mkApply__(gen.Select(t, defs.TUPLE_FIELD(2, 2)));
  }

  def SeqTrace_headState( arg: Tree  ):  Tree = { // REMOVE SeqTrace
    val t = gen.mkApply__(gen.Select(arg, defs.LIST_HEAD()));
    gen.mkApply__(gen.Select(t, defs.TUPLE_FIELD(2, 1)));

  }

  def SeqTrace_tail( arg: Tree ): Tree = { // REMOVE SeqTrace
    gen.mkApply__(gen.Select(arg, defs.LIST_TAIL()));
    }

    /** `<seqlist>.head()'
     */
  def SeqList_head( arg: Tree ): Tree = {
    gen.mkApply__(gen.Select(arg, defs.LIST_HEAD()));
  }

  def Negate(tree: Tree): Tree = {
    tree match {
      case Literal(BOOLEAN(value))=>
        gen.mkBooleanLit(tree.pos, !value);
      case _ =>
        gen.mkApply__(gen.Select(tree, defs.BOOLEAN_NOT()));
    }
  }

  /*protected*/ def And(left: Tree, right: Tree): Tree = {
    left match {
      case Literal(BOOLEAN(value)) =>
	return if(value) right else left;
      case _ =>
    }
    right match {
      case Literal(BOOLEAN(value)) =>
	if (value) return left;
      case _ =>
    }
    gen.mkApply_V(gen.Select(left, defs.BOOLEAN_AND()), Predef.Array[Tree](right));
  }

  /*protected*/ def Or(left: Tree, right: Tree): Tree = {
    left match {
      case Literal(BOOLEAN(value))=>
	    return if(value) left else right;
      case _ =>
    }
    right match {
      case Literal(BOOLEAN(value)) =>
	if (!value) return left;
      case _ =>
    }
    gen.mkApply_V(gen.Select(left, defs.BOOLEAN_OR()), Predef.Array[Tree](right));
  }

  // used by Equals
  private def getCoerceToInt(left: Type):  Symbol = {
    val sym = left.lookupNonPrivate(Names.coerce);
    //assert sym != Symbol.NONE : Debug.show(left);
    val syms = sym.alternativeSymbols();
    var i = 0; while( i < syms.length) {
      syms(i).info() match {
        case Type.MethodType(vparams, restpe) =>
          if (vparams.length == 0 && restpe.isSameAs(defs.INT_TYPE()))
            return syms(i);
        case _ =>
      }
      i = i + 1;
    }
    //assert false : Debug.show(left);
    throw new ApplicationError(left);
  }

  // used by Equals
  private def getEqEq(left: Type, right: Type): Symbol = {
    val sym = left.lookupNonPrivate(Names.EQEQ);
    //assert sym != Symbol.NONE
    //    : Debug.show(left) + "::" + Debug.show(left.members());
    var fun: Symbol  = null;
    var ftype:Type  = defs.ANY_TYPE();
    val syms = sym.alternativeSymbols();
    var  i = 0; while(i < syms.length) {
      val vparams = syms(i).valueParams();
      if (vparams.length == 1) {
        val vptype = vparams(0).info();
        if (right.isSubType(vptype) && vptype.isSubType(ftype)) {
          fun = syms(i);
          ftype = vptype;
        }
      }
      i = i + 1
    }
    //assert fun != null : Debug.show(sym.info());
    fun;
  }

  /*protected*/ def  Equals(left1:Tree , right1:Tree ): Tree = {
    var left = left1;
    var right = right1;
    val ltype = left.getType().widen();
    var rtype = right.getType().widen();
    if (ltype.isSameAs(rtype)
        && (ltype.isSameAs(defs.CHAR_TYPE())
            || ltype.isSameAs(defs.BYTE_TYPE())
            || ltype.isSameAs(defs.SHORT_TYPE())))
      {
        right = gen.mkApply__(gen.Select(right, getCoerceToInt(rtype)));
        rtype = defs.INT_TYPE();
      }
    val eqsym = getEqEq(ltype, rtype);
    gen.mkApply_V(gen.Select(left, eqsym), Predef.Array[Tree](right));
  }

  /*protected*/ def ThrowMatchError(pos: Int, tpe: Type ): Tree = {
    gen.mkApplyTV(
      gen.mkGlobalRef(pos, defs.MATCHERROR_FAIL()),
      Predef.Array[Tree](gen.mkType(pos, tpe)),
      Predef.Array[Tree](
        gen.mkStringLit(pos, unit.toString()),
        gen.mkIntLit(pos, Position.line(pos))
      ));
  }

  def ThrowMatchError(pos:int , tpe:Type , tree:Tree ):  Tree = {
    gen.mkApplyTV(
      gen.mkGlobalRef(pos, defs.MATCHERROR_REPORT()),
      Predef.Array[Tree](gen.mkType(pos, tpe)),
      Predef.Array[Tree](
        gen.mkStringLit(pos, unit.toString()),
        gen.mkIntLit(pos, Position.line(pos)),
        tree
      ));
  }

  def Error(pos: Int, tpe: Type): Tree = {
    gen.mkApplyTV(
      gen.mkGlobalRef(pos, defs.MATCHERROR_FAIL()),
      Predef.Array[Tree](gen.mkType(pos, tpe)),
      Predef.Array[Tree](
        gen.mkStringLit(pos, unit.toString()),
        gen.mkIntLit(pos, Position.line(pos))
      ));
  }


  def pairType(left: Type, right: Type): Type = {
    defs.TUPLE_TYPE(Predef.Array[Type] ( left, right ) );
  }

    def newPair(left: Tree, right: Tree): Tree = {
      gen.New(
        gen.mkApplyTV(
          gen.mkPrimaryConstructorGlobalRef( pos, defs.TUPLE_CLASS(2)),
          Predef.Array[Type] ( left.getType(), right.getType() ),
          Predef.Array[Tree] ( left, right )));
    }

}
}
